code/modules/reagents/chemistry/holder.dm

```diff
@@ -677,138 +727,325 @@
+	previous_reagent_list = list() //reset it to 0 - because any change will be different now.
+	update_total()
+	handle_reactions() //Should be okay without. Each step checks.
+
+/*
+* Force stops the current holder/reagents datum from reacting
+*
+* Calls end_reaction() for each equlilbrium datum in reaction_list and finish_reacting()
+* Usually only called when a datum is transfered into a NO_REACT container
+*/
+/datum/reagents/proc/force_stop_reacting()
+	var/mix_message = list()
+	for(var/e in reaction_list)
+		var/datum/equilibrium/E = e
+		mix_message += end_reaction(E)
+	if(length(mix_message))
+		my_atom.visible_message("<span class='notice'>[icon2html(my_atom, viewers(DEFAULT_MESSAGE_RANGE, src))] [mix_message.Join()]</span>")
+	finish_reacting()
+
+
+///Checks to see if the reagents has a difference in reagents_list and previous_reagent_list (I.e. if there's a difference between the previous call and the last)
+/datum/reagents/proc/has_changed_state()
+	var/total_matching_reagents = 0
+	for(var/R in previous_reagent_list)
+		if(has_reagent(R))
+			total_matching_reagents++
+	if(total_matching_reagents != reagent_list.len)
+		return TRUE
+	return FALSE
 
+/datum/reagents/proc/update_previous_reaction_list()
+	previous_reagent_list = list()
+	for(var/r in reagent_list)
+		var/datum/reagent/R = r
+		previous_reagent_list += R.type
+
+///Old reaction mechanics, edited to work on one only
+///This is changed from the old - purity of the reagents will affect yield
+/datum/reagents/proc/instant_react(datum/chemical_reaction/selected_reaction)
+	var/list/cached_required_reagents = selected_reaction.required_reagents
+	var/list/cached_results = selected_reaction.results
+	var/datum/cached_my_atom = my_atom
+	var/list/multiplier = INFINITY
+	for(var/B in cached_required_reagents)
+		multiplier = min(multiplier, round(get_reagent_amount(B) / cached_required_reagents[B]))
+
+	var/sum_purity = 0
+	for(var/B in cached_required_reagents)
+		var/datum/reagent/R = has_reagent(B)
+		remove_reagent(B, (multiplier * cached_required_reagents[B]), safety = 1)
+		sum_purity += R.purity
+	sum_purity /= cached_required_reagents.len
+
+	for(var/P in selected_reaction.results)
+		multiplier = max(multiplier, 1) //this shouldn't happen ...
+		var/yield = (cached_results[P]*multiplier)*sum_purity
+		SSblackbox.record_feedback("tally", "chemical_reaction", yield, P)
+		add_reagent(P, yield, null, chem_temp, sum_purity)
+
+	var/list/seen = viewers(4, get_turf(my_atom))
+	var/iconhtml = icon2html(cached_my_atom, seen)
+	if(cached_my_atom)
+		if(!ismob(cached_my_atom)) // No bubbling mobs
+			if(selected_reaction.mix_sound)
+				playsound(get_turf(cached_my_atom), selected_reaction.mix_sound, 80, TRUE)
+
+			for(var/mob/M in seen)
+				to_chat(M, "<span class='notice'>[iconhtml] [selected_reaction.mix_message]</span>")
+
+		if(istype(cached_my_atom, /obj/item/slime_extract))
+			var/obj/item/slime_extract/ME2 = my_atom
```

Comment:
If you could use a better variable than `ME2`, I'd appreciate it.
Prefer informative vars to this kind of cryptic few letters when possible.

Reply:
Sure, I've/I'll change it to extract